library/BerlinPipeline/src/Juvix/BerlinPipeline/Step/Types.hs

```diff
@@ -0,0 +1,48 @@
+module Juvix.BerlinPipeline.Step.Types
+  ( Env,
+    Named,
+    WorkingEnv,
+    ComputationalInput,
+    StepMetaInfo,
+    Interface,
+    register,
+  )
+where
+
+import qualified Juvix.BerlinPipeline.Meta as Meta
+import qualified Juvix.Sexp as Sexp
+import qualified Juvix.StandardLibrary.NameSymbol as NameSymbol
+
+data Env
+  = InContext NameSymbol.T
+  | SExp Sexp
+
+data T compin env where
+  Step ::
+    ( ComputationalInput compin,
+      WorkingEnv env
+    ) =>
+    compin ->
+    ComputationalOutput env
+
+data Named = NameStep
+  { name :: NameSymbol.T,
+    step :: forall compin env compout. T compin env
+  }
+
+class WorkingEnv (a :: * -> *) where
+  currentExp :: [Env]
+  context :: Context.T Sexp.T
+
+class ComputationalInput a where
+  languageData :: WorkingEnv a
+
+class StepMetaInfo a where
+  name :: Maybe NameSymbol.T
+  meta :: Meta.T
```

Comment:
I think these belong in `Juvix.Pipeline` and without the generics

Reply:
Without typeclasses, they will use the same types, that would make the structure too rigid.

Reply:
Yes, they will have to be the same type. If we look at the type of `WorkingEnv` it has two fixed contents a list of `EnvOrSexp` which contains either the S-expression itself, or a reference to the object in the `Context`. And the context itself, fixed to only hold `Sexpressions`

Since `ComputationalInput` only contains this and the `SurroundingEnvironment`, which is also fixed to have consistent data to flow to each step of the pass, it will also be fixed.

Reply:
(I guess it's fine if we have an `OO` style overloading of data and it's presentation, but I'm unsure of what it buys us here. I think that is my main issue?)